compiler/rustc_infer/src/infer/error_reporting/suggest.rs

```diff
@@ -351,6 +351,73 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
         }
     }
 
+    pub(super) fn suggest_function_pointers(
```

Comment:
This might be better named something like `note_and_explain_fn_pointers` or something, since you're doing a bit more than *just* a suggestion. That's a tiny nitpick though.